bootstrap/controllers/machine_deployment_controller.go

```diff
@@ -0,0 +1,437 @@
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get MachineDeployment: %w", err)
+	}
+
+	if !machineDeployment.DeletionTimestamp.IsZero() {
+		log.Info("MachineDeployment is being deleted, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	ownedMachines, err := r.getOwnedMachines(ctx, machineDeployment)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get owned machines: %w", err)
+	}
+
+	scope, err := r.createScope(machineDeployment, ownedMachines)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create scope: %w", err)
+	}
+
+	if !r.hasUpgradeInstructions(scope) {
+		log.Info("MachineDeployment has no upgrade instructions, removing upgrade-to annotation from machines and stopping reconciliation")
+		if err := r.removeUpgradeToFromMachines(ctx, log, scope); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to remove upgrade-to annotation from machines: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Starting the upgrade process
+	var upgradedMachines int
+	for _, m := range ownedMachines {
+		if r.isMachineUpgraded(scope, m) {
+			log.Info("Machine is already upgraded", "machine", m.Name)
+			upgradedMachines++
+			continue
+		}
+
+		if !m.DeletionTimestamp.IsZero() {
+			if !m.DeletionTimestamp.IsZero() {
+				log.Info("Machine is being deleted", "machine", m.Name)
+			}
+			continue
+		}
+
+		if r.isMachineUpradeFailed(m) {
+			log.Info("Machine upgrade failed for machine, requeuing...", "machine", m.Name)
+			if err := r.markUpgradeFailed(ctx, scope, m); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to mark upgrade as failed: %w", err)
+			}
+
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+
+		if r.isMachineUpgrading(m) {
+			log.Info("Machine is upgrading, requeuing...", "machine", m.Name)
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+
+		// Machine is not upgraded, mark it for upgrade
+		if err := r.markMachineToUpgrade(ctx, scope, m); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to mark machine to upgrade: %w", err)
+		}
+
+		log.Info("Machine marked for upgrade", "machine", m.Name)
+
+		if err := r.markUpgradeInProgress(ctx, scope); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to mark upgrade as in-progress: %w", err)
+		}
+	}
+
+	if upgradedMachines == len(ownedMachines) {
+		if err := r.markUpgradeDone(ctx, scope); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to mark upgrade as done: %w", err)
+		}
+
+		log.Info("All machines are upgraded")
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+}
+
+// markUpgradeInProgress marks the MachineDeployment as in-place upgrade in-progress.
+func (r *MachineDeploymentReconciler) markUpgradeInProgress(ctx context.Context, scope *MachineDeploymentUpgradeScope) error {
+	ant := scope.MachineDeployment.Annotations
```

Comment:
please just use the full word. There is no need for a reader to keep this (IMHO non-obvious) abbreviation in his mental model.